Resolving proxy from env on redirect

mbargiel opened the pull request:
The current `http` adapter implementation evaluates the proxy environment variables only once, when the request is initiated. When redirects are followed, the proxy is not re-resolved, which can be problematic in some cases. For instance:

* With `http_proxy` and `https_proxy` pointing to two different server hosts/ports, an `http:` request redirecting to an `https:` endpoint (or vice versa... go figure why) would result in the wrong proxy configuration used. (`curl` properly re-evaluates the environment variables on redirect and supports this use case.)
* With `no_proxy` set to one or more hosts, a request from one of the excluded hosts redirecting to an endpoint served by a non-excluded host would not use a proxy (when it should); likewise, a request from a non-excluded host redirecting to an endpoint served by an excluded host would use a proxy (when it shouldn't)

This PR implements a change in the `setProxy` code to always re-evaluate the proxy from environment, but only when the original request configuration did not include a proxy (ie. not changing the rules where "resolve from environment" kicks in). 

Also as part of this PR is a minor fix to allow using e.g. `http_proxy=https://proxyhost:proxyport`, so this PR should fix https://github.com/axios/axios/issues/3903. Basic HTTP proxying should be do-able over an encrypted (TLS) connection to the proxy, for instance when proxy credentials are used. Arguably, that's a strange use case, but there's no reason to not support it, especially when it's trivial to do so (`curl` does support using `http` endpoints over an `https` proxy.)

Ref: proposed contribution 1) mentioned in my comment on [issue #4318](https://github.com/axios/axios/issues/4318#issuecomment-1026089923).

mbargiel commented on lib/adapters/http.js:
```diff
@@ -13,33 +14,54 @@ var zlib = require('zlib');
 var enhanceError = require('../core/enhanceError');
-var defaults = require('../defaults');
 var Cancel = require('../cancel/Cancel');
+var defaults = require('../defaults');
```
Very minor change, I could revert it if you wish: this fixes a circular dependency issue that doesn't show up here, but which reared its head when I tried adding `var getProxyForUrl = require('proxy-from-env').getProxyForUrl;` after it. In the end, I added it above following alphabetical order... but please feel free to ask me to make changes to the `require` ordering.

mbargiel replied:
Looks like it's taken care of in https://github.com/axios/axios/pull/4220 in a much better way. I'll revert moving this require to avoid unnecessary merge conflicts .

mbargiel commented on lib/adapters/http.js:
```diff
@@ -13,33 +14,54 @@ var zlib = require('zlib');
+    if (proxyUrl) {
+      proxy = url.parse(proxyUrl);
+      // replace 'host' since the proxy object is not a URL object
+      proxy.host = proxy.hostname;
```
Avoids issues treating URL objects like an Axios proxy object. (Like `proxy.host` having a port).

mbargiel commented on lib/adapters/http.js:
```diff
@@ -152,82 +171,31 @@ module.exports = function httpAdapter(config) {
-        var shouldProxy = true;
-
-        if (noProxyEnv) {
-          var noProxy = noProxyEnv.split(',').map(function trim(s) {
-            return s.trim();
-          });
-
-          shouldProxy = !noProxy.some(function proxyMatch(proxyElement) {
-            if (!proxyElement) {
-              return false;
-            }
-            if (proxyElement === '*') {
-              return true;
-            }
-            if (proxyElement[0] === '.' &&
-                parsed.hostname.substr(parsed.hostname.length - proxyElement.length) === proxyElement) {
-              return true;
-            }
-
-            return parsed.hostname === proxyElement;
-          });
-        }
-
-        if (shouldProxy) {
-          proxy = {
-            host: parsedProxyUrl.hostname,
-            port: parsedProxyUrl.port,
-            protocol: parsedProxyUrl.protocol
-          };
-
-          if (parsedProxyUrl.auth) {
-            var proxyUrlAuth = parsedProxyUrl.auth.split(':');
-            proxy.auth = {
-              username: proxyUrlAuth[0],
-              password: proxyUrlAuth[1]
-            };
-          }
-        }
-      }
-    }
```
All this logic was merged into `setProxy`. This is the key to re-evaluating the environment variables on redirects.

Oh, ... and it was replaced with a single call to the `proxy-from-env` library (a [single line](https://github.com/axios/axios/pull/4436/files?w=1#diff-586c04c24608b86a074073c5514cc195be5c4c87ddb56da233f6d998ada4159fR31)).

mbargiel replied:
I noticed this could conflict with another PR that aims to improve support for NO_PROXY, though as far as I can see, the change they are trying to provide is already implemented in `proxy-from-env`.

sebinsua replied:
I think this incidentaly fixes another bug on line 189 as previously [`parsed.hostname` was assumed to be a `string` even though it could be `null`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v16/url.d.ts#L33).

jasonsaayman replied:
👍

mbargiel commented on lib/adapters/http.js:
```diff
@@ -152,82 +171,31 @@ module.exports = function httpAdapter(config) {
 
     var transport;
-    var isHttpsProxy = isHttpsRequest && (proxy ? isHttps.test(proxy.protocol) : true);
+    var isHttpsRequest = isHttps.test(options.protocol);
```
To use the right protocol defined by environment variable-set proxies on redirect, `setProxy` now overrides the `options.protocol` when needed. That allows us to simplify this logic a bit, but requires moving the `options.agent` assignment to after we set the proxy.

By the way, is it me or is the `agent` only necessary when `maxRedirects` is set to 0? Unless I'm mistaken, `httpsFollow/httpFollow` use `agents.https`/`agents.http`.

jasonsaayman replied:
👍

mbargiel commented on test/unit/adapters/http.js:
```diff
@@ -670,7 +670,7 @@ describe('supports http with nodejs', function () {
             host: 'localhost',
             port: 4000,
-            protocol: 'https'
+            protocol: 'https:'
```
If it's necessary to support both with and without trailing `:`, a small change is needed in `setProxy` to append it if it's missing (when `options.protocol` is overridden), otherwise it fails with an error in `follow-redirects`.

mbargiel commented on test/unit/adapters/http.js:
```diff
@@ -789,6 +789,56 @@ describe('supports http with nodejs', function () {
+    }).listen(4444, function () {
+      proxy = http.createServer(function (request, response) {
+        var parsed = url.parse(request.url);
+        if (parsed.pathname === '/redirected') {
```
I (somewhat hackily) re-used the proxy as both a proxy and the redirected endpoint. That saved me from starting another server to handle the redirection and having to deal with that extra server in `afterEach`. It would be cleaner, but that's for a single test. Let me know what you prefer.

jasonsaayman replied:
I think the single test is fine for now 👍

mbargiel commented on lib/adapters/http.js:
```diff
@@ -13,33 +14,54 @@ var zlib = require('zlib');
-  options.beforeRedirect = function beforeRedirect(redirection) {
-    redirection.headers.host = redirection.host;
-    setProxy(redirection, proxy, redirection.href);
+    options.headers.host = options.hostname + (options.port ? ':' + options.port : '');
```
This `headers.host` assignment was done two different ways:
- When initializing the request: https://github.com/axios/axios/pull/4436/files?w=1#diff-586c04c24608b86a074073c5514cc195be5c4c87ddb56da233f6d998ada4159fL216
- When following redirects: https://github.com/axios/axios/pull/4436/files?w=1#diff-586c04c24608b86a074073c5514cc195be5c4c87ddb56da233f6d998ada4159fL41

I moved this here because it's now possible to have different effective proxying on redirects when re-resolving the proxy from environment variables - ie set for `http_proxy` but not set for `https_proxy`, or skipping due to `no_proxy` for one call but not a redirect, etc.

jasonsaayman replied:
Cool 😄

mbargiel reviewed (commented):
Review pointers

mbargiel commented:
@jasonsaayman As requested, here's a PR that's fixing some protocol/proxy resolution issues. I thought it would be easier to review this first, then the HTTP CONNECT (https-proxy-agent) PR second. If you would rather have it all in a single "Fix proxy support" PR, let me know.

hrimhari commented:
Hi @jasonsaayman , is there anything we can do to help this move forward?

mbargiel commented:
I checked the conflicts, and they are trivial to solve, although one of them is worrying: I commented on the PR introducing the change (exposing `beforeRedirect` as a user option :+1:) because it can break proxy support on redirection (:-1:). I could solve the problem in this PR, by making sure we register a callback that in turns calls either a callback re-applying `setProxy` (when proxy is used), either the user callback (when `options.beforeRedirect` is specified), or both (when both are present/used).

11tonytony11 commented:
When will this fix be integrated to AXIOS?

jasonsaayman commented:
I will get to this one as soon as possible, I am trying to wrap up a v1 so that we can move forward with some large upgrades

mbargiel commented:
@jasonsaayman Thanks for the update! I'll sit tight until then. Let me know when is a good time to merge into my branch and fix conflicts. Also, how you want me to address the `beforeRedirect` option proxy regression I mentioned. I still have a follow-up PR, to actually integrate HTTP CONNECT into Axios, but I wanted to get this preliminary/orthogonal work out of the way first.

jasonsaayman commented:
@mbargiel please cna you do the merge conflict fixes, I will wait for this and then merge. This is a great and helpful contribution thanks 💯 🥇

mbargiel commented:
@jasonsaayman Yes, certainly! I'll try to fit this in my schedule today.

mbargiel commented:
@jasonsaayman I merged the main branch and added an extra commit to fix the regression introduced by the `config.beforeRedirect` support - see [1b4ca6d](https://github.com/axios/axios/pull/4436/commits/1b4ca6da5ab389f9b919388c227234438fd02cad).

I added a test with which I confirmed the suspected regression and validated the fix.

mbargiel commented:
If you prefer the latter `beforeRedirect` fix to be in its own, follow-up PR, I will happily oblige 😄

mbargiel commented:
Once this is merged, I'll get to work preparing the second half of my proxy fix contribution: supporting secure HTTPS proxying (over a TCP tunnel established using HTTP CONNECT). Note that I've been using [https-proxy-agent](github.com/TooTallNate/https-proxy-agent) but there are a few problems with it and I had to implement dirty workarounds - not fun. It's been suggested to me that [hpagent](https://www.npmjs.com/package/hpagent) might be a better contender so I'll look into that instead, but it might take me a whilie. Or I can share what I have that works with https-proxy-agent, and provide an hpagent replacement PR later.

mbargiel commented:
@jasonsaayman I'm not sure my `beforeRedirect` fix is good. My test only covers a single redirection, but I should verify the same case for more than 1 redirect (I suspect my code will append the beforeRedirect callback multiple times and/or the `config.beforeRedirect` will not be re-registered). If you don't mind, I'll pull it out of this PR and include it as a separate PR.

mbargiel commented:
See Issue https://github.com/axios/axios/issues/4703. PR to fix it will follow after this one is merged.

jasonsaayman commented:
Thanks very much for this 👍
